pyomo/contrib/alternative_solutions/solnpool.py

```diff
@@ -9,119 +9,1004 @@
+
+    @property
+    def last_solution(self):
+        """
+        Property to return last (successfully) added solution.
+        """
+        index = next(reversed(self._solutions.keys()))
+        return self._solutions[index]
+
+    @property
+    def policy(self):
+        """
+        Property to return pool construction policy.
+        """
+        return self.metadata['policy']
+
+    @property
+    def as_solution(self):
+        """
+        Property to return solution conversion method.
+        """
+        return self._as_solution
+
+    @property
+    def pool_config(self):
+        """
+        Property to return SolutionPool class specific configuration data.
+        """
+        return dict()
+
+    def __iter__(self):
+        for soln in self._solutions.values():
+            yield soln
+
+    def __len__(self):
+        return len(self._solutions)
+
+    def __getitem__(self, soln_id):
+        return self._solutions[soln_id]
+
+    def _next_solution_counter(self):
+        tmp = self.counter.solution_counter
+        self.counter.solution_counter += 1
+        return tmp
+
+
+class SolutionPool_KeepAll(SolutionPoolBase):
+    """
+    A SolutionPool subclass to keep all added solutions.
+
+    This class is designed to integrate with the alternative_solution generation methods.
+    Additionally, groups of solution pools can be handled with the PoolManager class.
+
+    Parameters
+    ----------
+    name : String
+        String name to describe the pool.
+    as_solution : Function or None
+        Method for converting inputs into Solution objects.
+        A value of None will result in the default _as_solution function being used.
+    counter : PoolCounter or None
+        PoolCounter object to manage solution indexing.
+        A value of None will result in a new PoolCounter object being created and used.
+    """
+
+    def __init__(self, name=None, as_solution=None, counter=None):
+        super().__init__(name, as_solution, counter, policy="keep_all")
+
+    def add(self, *args, **kwargs):
+        """
+        Add inputted solution to SolutionPool.
+        Relies on the instance as_solution conversion method to convert inputs to Solution Object.
+        Adds the converted Solution object to the pool dictionary.
+        ID value for the solution generated as next increment of instance PoolCounter.
+
+        Parameters
+        ----------
+        Input needs to match as_solution format from pool initialization.
+
+        Returns
+        ----------
+        int
+            The ID value to match the added solution from the solution pool's PoolCounter.
+            The ID value is also the pool dictionary key for this solution.
+        """
+        soln = self._as_solution(*args, **kwargs)
+        #
+        soln.id = self._next_solution_counter()
+        assert (
+            soln.id not in self._solutions
+        ), f"Solution id {soln.id} already in solution pool context '{self._context_name}'"
```

Comment:
Perhaps make this a `DeveloperError`? This comment applies to a few of the other assertions in this PR too.